### DISCUSSION TITLE
Background tasks in tests don't work properly

### DISCUSSION THREAD

**Discussion by pmiddend** [fastapi/fastapi#10975] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# File "bug.py"

import asyncio
from time import time

from fastapi import BackgroundTasks
from fastapi import FastAPI

app = FastAPI()


async def some_coroutine() -> None:
    print(f"{time()} sleeping start")
    await asyncio.sleep(3.0)
    print(f"{time()} sleeping end")


@app.put("/")
async def detector_arm(
    background_tasks: BackgroundTasks,
) -> str:
    background_tasks.add_task(some_coroutine)
    print(f"{time()} returning")
    return "1"

# File test_bug.py

from time import time

from fastapi.testclient import TestClient

from simplon_stub.bug import app


def test_bug() -> None:
    with TestClient(app) as test_client:
        print(f"{time()} starting request")
        response_put = test_client.put("/")
        print(f"{time()} request done")
        assert response_put.status_code == 200
        assert False
```

### Description

I'm running the test. I'm expecting the background task to be executed in the background and the `PUT` request to return immediately. However, the log looks like this:

```
1705389741.1490555 starting request
1705389741.1499116 returning
1705389741.1499949 sleeping start
1705389744.1537182 sleeping end
1705389744.155864 request done
```

Meaning:

- the put handler is done (see the line with `returning`)
- the test resumes after the background task completed, instead of immediately

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

1.10.13

### Python Version

Python 3.10.12

**Comment by Kludex** [comment, marked as answer]
It's the intended behavior. I think what you want was the behavior some years ago, if I'm not mistaken (I may be).

In any case, I think the current behavior is fine.

**Reply by pmiddend** [reply]
The intended behavior is that a background task is not running in the background? That's at least a little unexpected I would say.

Anyway, is there a "fix" for this? I would like to have something running in the background that is "triggered" by a HTTP request, and would like to write tests using this background logic.

**Reply by thentgesMindee** [reply]
@Kludex isn't it a bit weird to have this behavior, now that ([since 0.106 release](https://github.com/tiangolo/fastapi/releases/tag/0.106.0)) background tasks can't use resources from the actual endpoint ?

I mean, before 0.106, I don't see any issue having the background tasks not running in background, since anyway they'll have access to the same objects whether it was in background or not.

Now, with 0.106 applied, we can have a situation where a unit test validate some logic happening in a background task (using for example the same db session than in the endpoint) that works, whereas outside of tests, it would break since the session would be closed before entering the background task.

If this is really the expected behavior, how would you recommend testing such logic, to make sure the background tasks are not using already closed objects ?

**Reply by thentgesMindee** [reply]
Sorry to be a bit pushy, but I feel like the question might have went under the radar @Kludex , any recommendations on that ?
EDIT: maybe I should create an issue?
